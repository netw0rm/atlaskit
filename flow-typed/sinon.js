declare module "sinon" {
  // Telling flow to ignore the typing for now
  declare export default any;
}
