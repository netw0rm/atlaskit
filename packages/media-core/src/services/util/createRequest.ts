import axios from 'axios';
import {MediaApiConfig} from '../../config';
import {checkWebpSupport} from '../../utils';

export type ResponseType = 'json' | 'image';
export interface CreateRequestFunc {
  (requestOptions: RequestOptions): Promise<any>;
}

export interface RequesterOptions {
  clientId: string;
  collectionName?: string;
  config: MediaApiConfig;
}

export interface RequestOptions {
  method?: string;
  url: string;
  params?: Object;
  headers?: Object;
  data?: Object;
  responseType?: ResponseType;
}

const buildHeaders = (requesterOptions: RequesterOptions, requestOptions: RequestOptions, token: string) => {
  const headers = {
    ...requestOptions.headers,
    'X-Client-Id': requesterOptions.clientId,
    'Authorization': `Bearer ${token}`,
    'Content-Type': 'application/json'
  } as any;

  if (requestOptions.responseType === 'image') {
    return checkWebpSupport().then(isWebpSupported => {
      // q=0.8 stands for 'quality factor' => http://stackoverflow.com/a/10496722
      const noWebpAcceptHeader = 'image/*,*/*;q=0.8';
      const webpAcceptHeader = 'image/webp,image/*,*/*;q=0.8';

      headers.accept = isWebpSupported ? webpAcceptHeader : noWebpAcceptHeader;
      return headers;
    });
  }

  return Promise.resolve(headers);
};

const responseTypeToAxios = (responseType?: ResponseType): string => {
  responseType = responseType || 'json';

  const responseTypeMap = {
    image: 'blob',
    json: 'json'
  };

  return responseTypeMap[responseType];
};

export default (requesterOptions: RequesterOptions) => (requestOptions: RequestOptions): any => {
  return requesterOptions.config.tokenProvider(requesterOptions.collectionName).then(token => {
    return buildHeaders(requesterOptions, requestOptions, token).then(headers => {
      const responseType = responseTypeToAxios(requestOptions.responseType);
      const {method, url, params, data} = requestOptions;
      const {config, collectionName} = requesterOptions;

      return axios({
        method: method || 'get',
        url: url,
        baseURL: config.serviceHost,
        headers,
        params: {
          collection: collectionName,
          ...params
        },
        data: data,
        responseType
      }).then(response => response.data);
    });
  });
};
