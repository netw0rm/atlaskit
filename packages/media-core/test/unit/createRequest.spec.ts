import * as sinon from 'sinon';
import createRequest from '../../src/services/util/createRequest';

let fakeXhr;
let myFakeResponses;

describe('createRequest()', () => {
  const token = 'ABC';
  const clientId = '1234';
  const serviceHost = 'http://example.com';

  beforeEach(() => {
    myFakeResponses = [];
    fakeXhr = sinon.useFakeXMLHttpRequest();
    fakeXhr.onCreate = res => myFakeResponses.push(res);
  });

  afterEach(() => {
    fakeXhr.restore();
  });

  it('should send the client ID and auth token in header fields by default', () => {
    const tokenProvider = sinon.stub().returns(Promise.resolve(token));

    const request = createRequest({
      clientId,
      config: {
        tokenProvider,
        serviceHost
      }
    });

    setTimeout(() => myFakeResponses[0].respond(200), 0);

    return request({url: '/some-api/links'}).then(json => {
      expect(tokenProvider.calledOnce).toBe(true);
      expect(myFakeResponses[0].url).toBe('http://example.com/some-api/links');
      expect(myFakeResponses[0].requestHeaders['X-Client-Id']).toBe(clientId);
      expect(myFakeResponses[0].requestHeaders['Authorization']).toBe(`Bearer ${token}`);
    });
  });

  it('should send auth arguments using queryParams when preventPreflight is true', () => {
    const tokenProvider = sinon.stub().returns(Promise.resolve(token));

    const request = createRequest({
      clientId,
      config: {
        tokenProvider,
        serviceHost
      },
      preventPreflight: true
    });

    setTimeout(() => myFakeResponses[0].respond(200), 0);

    return request({url: '/some-api/links'}).then(json => {
      expect(tokenProvider.calledOnce).toBe(true);
      expect(myFakeResponses[0].url).toBe(`http://example.com/some-api/links?token=${token}&client=${clientId}`);
      expect(myFakeResponses[0].requestHeaders['X-Client-Id']).toBeUndefined();
      expect(myFakeResponses[0].requestHeaders['Authorization']).toBeUndefined();
    });
  });
});
