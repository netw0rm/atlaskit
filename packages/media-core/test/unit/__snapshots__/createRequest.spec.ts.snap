// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRequest() should send auth arguments using queryParams when preventPreflight is true 1`] = `"http://example.com/some-api/links?token=ABC&client=1234"`;
